src/newgrp.c: sg(1): Various source code improvements and bug fixes

alejandro-colomar opened the pull request:
Cc: @stoeckmann , @ikerexxe 

---

Revisions:

<details>
<summary>v2</summary>

-  Unname `argc`.

```
$ git rd -U0
 1:  ef0420019 =  1:  ef0420019 src/newgrp.c: Document '-l'
 2:  b6d4f9b7a =  2:  b6d4f9b7a src/newgrp.c: sg(1): Refactor, to remove 'else' branch
 3:  56166b669 =  3:  56166b669 src/newgrp.c: sg(1): -c: Fail if the option is used without an argument
 4:  2b8316a3f =  4:  2b8316a3f src/newgrp.c: Split conditional block
 5:  8046815f8 !  5:  6843c19b3 src/newgrp.c: Use argv instead of argc to check an argument
    @@ Commit message
    +    Unname 'argc' to avoid diagnostics about the unused argument.
    +
    @@ src/newgrp.c: static void syslog_sg (const char *name, const char *group)
    -+main(int argc, char *argv[argc + 1])
    ++main(int, char *argv[])
 6:  17c01e5f4 !  6:  028e5c748 src/newgrp.c: Check '-l' and '-' in the documented order
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
 7:  82c563b90 !  7:  3a0892de5 src/newgrp.c: newgrp(1): Remove 'else's after 'goto'
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
    @@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
 8:  e7ee9d344 !  8:  4246adf9a src/newgrp.c: De-duplicate check
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
    @@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
 9:  f58234eb3 !  9:  1f62687b7 src/newgrp.c: Separate code handling group name and code handling command
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
    @@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
    @@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
10:  03bb4b4ef ! 10:  67834ed31 src/newgrp.c: Refactor conditionals to de-duplicate code
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
    @@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
11:  5ae74951d ! 11:  f11890b4f src/newgrp.c: Fix indentation
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
12:  de4cc76cc ! 12:  85706b5ec src/newgrp.c: Fail if there are trailing (unexpected) arguments
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
13:  33dfdafe9 ! 13:  499eadeb2 src/newgrp.c: Remove incorrect comment
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
14:  a9af91f67 ! 14:  3573e24a6 src/newgrp.c: sg(1): Use 'goto failure;' to fail
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
15:  6b652d160 = 15:  d78ede7e7 lib/prototypes.h: Fix indentation
16:  99c712e35 ! 16:  89c89007c src/newgrp.c: Report SHADOW_AUDIT_FAILURE on failure
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
17:  41fa8b9e8 ! 17:  657823c6f src/newgrp.c: Fix error handling
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
18:  fc5e5462b = 18:  919b2c77d src/newgrp.c: syslog_sg(): Use Prog instead of its pattern
19:  23acca2c6 ! 19:  28aae3568 src/newgrp.c: Use idiomatic pointer arithmetic notation
    @@ src/newgrp.c
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
    @@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
    @@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
    @@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
20:  0e4f97b2a ! 20:  e4266db15 src/newgrp.c: Remove redundant comment
    @@ src/newgrp.c: static void check_perms (const struct group *grp,
    -@@ src/newgrp.c: main(int argc, char *argv[argc + 1])
    +@@ src/newgrp.c: main(int, char *argv[])
```
</details>

<details>
<summary>v3</summary>

-  Accept both `-l` and `-` in the same command, for consistency with su(1).  [@stoeckmann ]
-  Remove redundant input validation.  [@stoeckmann ]

```
$ git rd 
 1:  ef0420019 =  1:  ef0420019 src/newgrp.c: Document '-l'
 2:  b6d4f9b7a =  2:  b6d4f9b7a src/newgrp.c: sg(1): Refactor, to remove 'else' branch
 3:  56166b669 =  3:  56166b669 src/newgrp.c: sg(1): -c: Fail if the option is used without an argument
 4:  2b8316a3f =  4:  2b8316a3f src/newgrp.c: Split conditional block
 5:  6843c19b3 =  5:  6843c19b3 src/newgrp.c: Use argv instead of argc to check an argument
 6:  028e5c748 =  6:  028e5c748 src/newgrp.c: Check '-l' and '-' in the documented order
 7:  3a0892de5 =  7:  3a0892de5 src/newgrp.c: newgrp(1): Remove 'else's after 'goto'
 8:  4246adf9a =  8:  4246adf9a src/newgrp.c: De-duplicate check
 9:  1f62687b7 =  9:  1f62687b7 src/newgrp.c: Separate code handling group name and code handling command
10:  67834ed31 = 10:  67834ed31 src/newgrp.c: Refactor conditionals to de-duplicate code
11:  f11890b4f = 11:  f11890b4f src/newgrp.c: Fix indentation
12:  85706b5ec = 12:  85706b5ec src/newgrp.c: Fail if there are trailing (unexpected) arguments
13:  499eadeb2 = 13:  499eadeb2 src/newgrp.c: Remove incorrect comment
14:  3573e24a6 = 14:  3573e24a6 src/newgrp.c: sg(1): Use 'goto failure;' to fail
15:  d78ede7e7 = 15:  d78ede7e7 lib/prototypes.h: Fix indentation
16:  89c89007c = 16:  89c89007c src/newgrp.c: Report SHADOW_AUDIT_FAILURE on failure
17:  657823c6f = 17:  657823c6f src/newgrp.c: Fix error handling
18:  919b2c77d = 18:  919b2c77d src/newgrp.c: syslog_sg(): Use Prog instead of its pattern
19:  28aae3568 = 19:  28aae3568 src/newgrp.c: Use idiomatic pointer arithmetic notation
20:  e4266db15 = 20:  e4266db15 src/newgrp.c: Remove redundant comment
 -:  --------- > 21:  29c24fed6 src/newgrp.c: Accept both -l and - in the same command
 -:  --------- > 22:  df3bb4cbe src/newgrp.c: Remove redundant input validation
```
</details>

<details>
<summary>v3b</summary>

-  Rebase

```
$ git rd 
 1:  ef04200199d2 =  1:  5bc5e6157526 src/newgrp.c: Document '-l'
 2:  b6d4f9b7a5c8 =  2:  bf9e61c0ef7e src/newgrp.c: sg(1): Refactor, to remove 'else' branch
 3:  56166b6697a6 =  3:  fefe02ecb687 src/newgrp.c: sg(1): -c: Fail if the option is used without an argument
 4:  2b8316a3f78d =  4:  3635176e2822 src/newgrp.c: Split conditional block
 5:  6843c19b360d =  5:  3aa5a64e3a31 src/newgrp.c: Use argv instead of argc to check an argument
 6:  028e5c748b33 =  6:  4bf4e30960e3 src/newgrp.c: Check '-l' and '-' in the documented order
 7:  3a0892de52b6 =  7:  8b29b89fee14 src/newgrp.c: newgrp(1): Remove 'else's after 'goto'
 8:  4246adf9a5f4 =  8:  e007fbcb1eaa src/newgrp.c: De-duplicate check
 9:  1f62687b7ff4 =  9:  36f9a127dfae src/newgrp.c: Separate code handling group name and code handling command
10:  67834ed31be1 = 10:  c74525577028 src/newgrp.c: Refactor conditionals to de-duplicate code
11:  f11890b4f79a = 11:  e23d52c64070 src/newgrp.c: Fix indentation
12:  85706b5ec1be = 12:  4e0209d0d6af src/newgrp.c: Fail if there are trailing (unexpected) arguments
13:  499eadeb203d = 13:  deda72428f86 src/newgrp.c: Remove incorrect comment
14:  3573e24a6f39 = 14:  f9d2fadd0bf4 src/newgrp.c: sg(1): Use 'goto failure;' to fail
15:  d78ede7e76cd = 15:  b91bed51b80c lib/prototypes.h: Fix indentation
16:  89c89007c03c = 16:  7ceadd0af331 src/newgrp.c: Report SHADOW_AUDIT_FAILURE on failure
17:  657823c6f1ca = 17:  d34e55368aac src/newgrp.c: Fix error handling
18:  919b2c77d4ba = 18:  b526d9097b61 src/newgrp.c: syslog_sg(): Use Prog instead of its pattern
19:  28aae35684c0 = 19:  29a64c5b7339 src/newgrp.c: Use idiomatic pointer arithmetic notation
20:  e4266db158a6 = 20:  22ad719bfcbd src/newgrp.c: Remove redundant comment
21:  29c24fed60a0 = 21:  d08282fce523 src/newgrp.c: Accept both -l and - in the same command
22:  df3bb4cbea44 = 22:  9faab479f973 src/newgrp.c: Remove redundant input validation
```
</details>

<details>
<summary>v3c</summary>

-  Rebase

```
$ git rd 
 1:  5bc5e615 =  1:  59265b45 src/newgrp.c: Document '-l'
 2:  bf9e61c0 =  2:  f6f71954 src/newgrp.c: sg(1): Refactor, to remove 'else' branch
 3:  fefe02ec =  3:  de6ea3a7 src/newgrp.c: sg(1): -c: Fail if the option is used without an argument
 4:  3635176e =  4:  6ac361bd src/newgrp.c: Split conditional block
 5:  3aa5a64e =  5:  721a3d65 src/newgrp.c: Use argv instead of argc to check an argument
 6:  4bf4e309 =  6:  33a370f9 src/newgrp.c: Check '-l' and '-' in the documented order
 7:  8b29b89f !  7:  07747cce src/newgrp.c: newgrp(1): Remove 'else's after 'goto'
    @@ src/newgrp.c: main(int, char *argv[])
                                fprintf (
                                        stderr, _("%s: provided group is not a valid group name\n"),
     @@ src/newgrp.c: main(int, char *argv[])
    -                           SYSLOG ((LOG_CRIT, "GID '%lu' does not exist",
    -                                   (unsigned long) pwd->pw_gid));
    +                           SYSLOG(LOG_CRIT, "GID '%lu' does not exist",
    +                                  (unsigned long) pwd->pw_gid);
                                goto failure;
     -                  } else {
     -                          group = grp->gr_name;
 8:  e007fbcb =  8:  aa6e589f src/newgrp.c: De-duplicate check
 9:  36f9a127 =  9:  4a9d2c27 src/newgrp.c: Separate code handling group name and code handling command
10:  c7452557 = 10:  6cfa7edf src/newgrp.c: Refactor conditionals to de-duplicate code
11:  e23d52c6 ! 11:  0aa744cd src/newgrp.c: Fix indentation
    @@ src/newgrp.c: main(int, char *argv[])
     -                          fprintf (stderr,
     -                                   _("%s: GID '%lu' does not exist\n"),
     -                                   Prog, (unsigned long) pwd->pw_gid);
    --                          SYSLOG ((LOG_CRIT, "GID '%lu' does not exist",
    --                                  (unsigned long) pwd->pw_gid));
    +-                          SYSLOG(LOG_CRIT, "GID '%lu' does not exist",
    +-                                 (unsigned long) pwd->pw_gid);
     -                          goto failure;
     -                  }
     -                  group = grp->gr_name;
    @@ src/newgrp.c: main(int, char *argv[])
     +          if (NULL == grp) {
     +                  fprintf(stderr, _("%s: GID '%lu' does not exist\n"),
     +                          Prog, (unsigned long) pwd->pw_gid);
    -+                  SYSLOG((LOG_CRIT, "GID '%lu' does not exist",
    -+                         (unsigned long) pwd->pw_gid));
    ++                  SYSLOG(LOG_CRIT, "GID '%lu' does not exist",
    ++                         (unsigned long) pwd->pw_gid);
     +                  goto failure;
     +          }
     +          group = grp->gr_name;
12:  4e0209d0 = 12:  0e4f56bd src/newgrp.c: Fail if there are trailing (unexpected) arguments
13:  deda7242 = 13:  737b4246 src/newgrp.c: Remove incorrect comment
14:  f9d2fadd = 14:  0476fb9b src/newgrp.c: sg(1): Use 'goto failure;' to fail
15:  b91bed51 = 15:  0de44ac5 lib/prototypes.h: Fix indentation
16:  7ceadd0a = 16:  4d82e688 src/newgrp.c: Report SHADOW_AUDIT_FAILURE on failure
17:  d34e5536 = 17:  711ad17f src/newgrp.c: Fix error handling
18:  b526d909 = 18:  61867cb6 src/newgrp.c: syslog_sg(): Use Prog instead of its pattern
19:  29a64c5b = 19:  b2dfc346 src/newgrp.c: Use idiomatic pointer arithmetic notation
20:  22ad719b = 20:  9505b7b0 src/newgrp.c: Remove redundant comment
21:  d08282fc = 21:  5463f229 src/newgrp.c: Accept both -l and - in the same command
22:  9faab479 = 22:  aff2bd8b src/newgrp.c: Remove redundant input validation
```
</details>

<details>
<summary>v3d</summary>

-  Rebase

```
$ git rd 
 1:  59265b45 =  1:  461e4a09 src/newgrp.c: Document '-l'
 2:  f6f71954 =  2:  ff803096 src/newgrp.c: sg(1): Refactor, to remove 'else' branch
 3:  de6ea3a7 =  3:  ae4974a5 src/newgrp.c: sg(1): -c: Fail if the option is used without an argument
 4:  6ac361bd =  4:  b55d73b4 src/newgrp.c: Split conditional block
 5:  721a3d65 =  5:  acb6da11 src/newgrp.c: Use argv instead of argc to check an argument
 6:  33a370f9 =  6:  e07b06db src/newgrp.c: Check '-l' and '-' in the documented order
 7:  07747cce =  7:  f76e18f3 src/newgrp.c: newgrp(1): Remove 'else's after 'goto'
 8:  aa6e589f =  8:  b2142554 src/newgrp.c: De-duplicate check
 9:  4a9d2c27 =  9:  eb01dbc5 src/newgrp.c: Separate code handling group name and code handling command
10:  6cfa7edf = 10:  1df97c87 src/newgrp.c: Refactor conditionals to de-duplicate code
11:  0aa744cd = 11:  fc4443aa src/newgrp.c: Fix indentation
12:  0e4f56bd = 12:  3d964c57 src/newgrp.c: Fail if there are trailing (unexpected) arguments
13:  737b4246 = 13:  e7f6ca99 src/newgrp.c: Remove incorrect comment
14:  0476fb9b = 14:  ea162c9f src/newgrp.c: sg(1): Use 'goto failure;' to fail
15:  0de44ac5 = 15:  63805d1a lib/prototypes.h: Fix indentation
16:  4d82e688 = 16:  2aae9238 src/newgrp.c: Report SHADOW_AUDIT_FAILURE on failure
17:  711ad17f = 17:  1076a30c src/newgrp.c: Fix error handling
18:  61867cb6 = 18:  acccf15c src/newgrp.c: syslog_sg(): Use Prog instead of its pattern
19:  b2dfc346 = 19:  8ee07de8 src/newgrp.c: Use idiomatic pointer arithmetic notation
20:  9505b7b0 = 20:  68031eb3 src/newgrp.c: Remove redundant comment
21:  5463f229 = 21:  e170755c src/newgrp.c: Accept both -l and - in the same command
22:  aff2bd8b = 22:  f3559537 src/newgrp.c: Remove redundant input validation
```
</details>

<details>
<summary>v3e</summary>

-  Rebase

```
$ git rd 
 1:  461e4a098890 =  1:  08012a68b4d6 src/newgrp.c: Document '-l'
 2:  ff8030962706 =  2:  de307e311fb7 src/newgrp.c: sg(1): Refactor, to remove 'else' branch
 3:  ae4974a5578a =  3:  d65c8397d076 src/newgrp.c: sg(1): -c: Fail if the option is used without an argument
 4:  b55d73b4c09a =  4:  4b8b87b50b98 src/newgrp.c: Split conditional block
 5:  acb6da112132 =  5:  a8a865abda9c src/newgrp.c: Use argv instead of argc to check an argument
 6:  e07b06db83e3 =  6:  3e5706342fbe src/newgrp.c: Check '-l' and '-' in the documented order
 7:  f76e18f3698c =  7:  4088994dd0d5 src/newgrp.c: newgrp(1): Remove 'else's after 'goto'
 8:  b21425547c8e =  8:  c89ed108df7d src/newgrp.c: De-duplicate check
 9:  eb01dbc58edf =  9:  bea61ad4a0b2 src/newgrp.c: Separate code handling group name and code handling command
10:  1df97c87644e = 10:  1656c7a41426 src/newgrp.c: Refactor conditionals to de-duplicate code
11:  fc4443aa0169 = 11:  3a9bdee0e3bf src/newgrp.c: Fix indentation
12:  3d964c5758f5 = 12:  6472b1cac7ed src/newgrp.c: Fail if there are trailing (unexpected) arguments
13:  e7f6ca998151 = 13:  8381db795e98 src/newgrp.c: Remove incorrect comment
14:  ea162c9fee83 = 14:  0a975191f8ca src/newgrp.c: sg(1): Use 'goto failure;' to fail
15:  63805d1a2f48 = 15:  7ba423811286 lib/prototypes.h: Fix indentation
16:  2aae9238f3b1 = 16:  9a2be39d7175 src/newgrp.c: Report SHADOW_AUDIT_FAILURE on failure
17:  1076a30c0a14 = 17:  32bdced784c7 src/newgrp.c: Fix error handling
18:  acccf15ca5d1 = 18:  ca9e281e9cdd src/newgrp.c: syslog_sg(): Use Prog instead of its pattern
19:  8ee07de87ae7 = 19:  df675fc9faaf src/newgrp.c: Use idiomatic pointer arithmetic notation
20:  68031eb36c41 = 20:  703ff72f6f0f src/newgrp.c: Remove redundant comment
21:  e170755cd2f3 = 21:  9529daa92cac src/newgrp.c: Accept both -l and - in the same command
22:  f35595370cd8 = 22:  1fca9d93738f src/newgrp.c: Remove redundant input validation
```
</details>

alejandro-colomar committed:
src/newgrp.c: Document '-l'

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: sg(1): Refactor, to remove 'else' branch

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: sg(1): -c: Fail if the option is used without an argument

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Split conditional block

This way, the code is better self-documented, in the way that '-c' is
pretty much independent of the actual command, except that if it is
specified, a command is required, but is otherwise skipped when parsing
the command line.

It also reduces indentation, which is another way of saying it reduces
complexity of the code and thus improves readability.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Use argv instead of argc to check an argument

It does the same, and avoids having to update both argc and argv, which
is prone to mistakes.

Unname 'argc' to avoid diagnostics about the unused argument.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Check '-l' and '-' in the documented order

This makes the code more readable, by more closely matching its
documentation.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: newgrp(1): Remove 'else's after 'goto'

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: De-duplicate check

Both sg(1) and newgrp(1) reject any flags after the optional '-l'/'-'.
Do the check once in the common path.

Also, remove some comments that are getting less correct.

BTW, in one path we were doing the check with manual byte operations,
and in the other with strprefix().  Let's unify with neither of them;
strspn(3) is standard, and better (it allows multi-character search in
a single call, which we need elsewhere, which is why I'll unify the
entire project with strspn(3) in a future commit).

Also BTW, in the case of sg(1), we were exiting early, without
'goto failure;'.  I presume that was an accident.  The only difference
is the audit code, which I don't see a reason to skip.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Separate code handling group name and code handling command

This will allow some deduplication of code in a following commit.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Refactor conditionals to de-duplicate code

Both sg(1) and newgrp(1) handle the group name in the same way if it
exists.  They differ in how they handle the absence of it.  Reflect that
in the source code.

Use a block to avoid changing the indentation.  This will make the diff
easier to review.  A future patch will clean up the indentation.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Fix indentation

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Fail if there are trailing (unexpected) arguments

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Remove incorrect comment

This comment has several incorrections:

-  '-l'/'-' is now exclusive of newgrp(1); it works in sg(1) too.
-  The group name is mandatory in sg(1).
-  'Any remaining arguments' are not used to execute a command.
   Only the first one is used, and everything else was silently
   being ignored, until in last commit I made it fail hard.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: sg(1): Use 'goto failure;' to fail

Everywhere else we were failing that way.  I don't see any reason to
avoid the audit code here.  It probably was a mistake.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
lib/prototypes.h: Fix indentation

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Report SHADOW_AUDIT_FAILURE on failure

Also, align the arguments more consistently with the call above.

Fixes: 8dfe21f592e9 (2025-03-03; "src/: update group audit messages")
Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Fix error handling

Use 'goto failure;', as we forgot to closelog(3).

Also, as always, remove the redundant nearby comment.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: syslog_sg(): Use Prog instead of its pattern

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Use idiomatic pointer arithmetic notation

Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Remove redundant comment

The usage() function already documents the syntax.  Let's keep a single
source of truth.

Signed-off-by: Alejandro Colomar <alx@kernel.org>

stoeckmann commented on src/newgrp.c:
```diff
@@ -64,16 +64,14 @@ static void check_perms (const struct group *grp,
+	if (is_newgrp)
+		fputs(_("Usage: newgrp [-l|-] [group]\n"), stderr);
+	else
+		fputs(_("Usage: sg [-l|-] group [[-c] command]\n"), stderr);
```
Diverges from `su`'s usage message, which separates `[-]` from the other options. It also just states `[options]`, something maintainers should decide about what they actually want their tools to look like.

alejandro-colomar replied:
su(1) actually behaves differently.


`su -l -` is correct, as it first parses options with getopt_long(3) and later parses an optional `-`

<https://github.com/shadow-maint/shadow/blob/47be361b863f550bda561abcac51fb125218b19f/src/su.c#L831>
<https://github.com/shadow-maint/shadow/blob/47be361b863f550bda561abcac51fb125218b19f/src/su.c#L840-L842>
<https://github.com/shadow-maint/shadow/blob/47be361b863f550bda561abcac51fb125218b19f/src/su.c#L859-L862>

But `sg -l -` is not correct, as it only accepts one of `-l` or `-` but not both.

<https://github.com/shadow-maint/shadow/blob/47be361b863f550bda561abcac51fb125218b19f/src/newgrp.c#L471-L477>

We may want to align the behavior of sg(1) with that of su(1), but I'm currently documenting how they do actually behave.

I'll add a patch aligning the behavior.  I'll also align newgrp(1) with su(1), just to keep sg(1) and newgrp(1) as close as possible.

Thanks for pointing out this inconsistency!

stoeckmann commented on src/newgrp.c:
```diff
@@ -451,120 +447,57 @@ int main (int argc, char **argv)
+	if (*argv != NULL && strspn(*argv, "-")) {
+		usage();
+		goto failure;
+	}
```
This whole block can be removed, since this check is already covered in `is_valid_group_name`.

stoeckmann reviewed (commented):
Whatever. If shadow devs break yet another setuid binary, I'm here to call them out.

alejandro-colomar commented on src/newgrp.c:
```diff
@@ -367,7 +365,8 @@ static void syslog_sg (const char *name, const char *group)
  */
-int main (int argc, char **argv)
+int
+main(int argc, char *argv[argc + 1])
```
This is supported since ANSI C89.

alejandro-colomar replied:
Fixed in v2.

alejandro-colomar committed:
src/newgrp.c: Accept both -l and - in the same command

This is for consistency with su(1).

Reported-by: Tobias Stoeckmann <tobias@stoeckmann.org>
Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar committed:
src/newgrp.c: Remove redundant input validation

is_valid_group_name() already fails if the name starts with '-'.
This changes the diagnostic message, but other than that it's fine.
Let's prefer the simpler code.

Reported-by: Tobias Stoeckmann <tobias@stoeckmann.org>
Suggested-by: Tobias Stoeckmann <tobias@stoeckmann.org>
Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar commented on src/newgrp.c:
```diff
@@ -451,120 +447,57 @@ int main (int argc, char **argv)
-		}
-#endif
-		exit(EXIT_FAILURE);
+		goto failure;
```
No problem; thanks for confirming.

alejandro-colomar commented on src/newgrp.c:
```diff
@@ -451,120 +447,57 @@ int main (int argc, char **argv)
+	if (*argv != NULL) {
+		if (!is_valid_group_name(*argv)) {
+			fprintf(stderr, _("%s: provided group is not a valid group name\n"),
+			        Prog);
```
Thanks!
